src/main/scala/tech/sourced/gemini/ConnectedComponents.scala

```diff
@@ -0,0 +1,125 @@
+package tech.sourced.gemini
+
+import java.nio.ByteBuffer
+
+import com.datastax.driver.core.{Session, SimpleStatement}
+import org.slf4j.{Logger => Slf4jLogger}
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
+
+case class FileHash(sha1: String, value: ByteBuffer)
+
+class DBConnectedComponents(
+                              log: Slf4jLogger,
+                              conn: Session,
+                              keyspace: String = Gemini.defautKeyspace) extends ConnectedComponents(log) {
+  def getHashtables(): List[Byte] = {
+    val hashtablesCql = s"SELECT DISTINCT hashtable FROM $keyspace.hashtables"
+
+    conn
+      .execute(new SimpleStatement(hashtablesCql))
+      .asScala
+      .map(_.getByte("hashtable"))
+      .toList.
+      sorted
+  }
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash] = {
+    val cql = s"SELECT sha1, value FROM $keyspace.hashtables WHERE hashtable=$hashtable"
+
+    conn.execute(new SimpleStatement(cql))
+      .asScala
+      .map(row => FileHash(row.getString("sha1"), row.getBytes("value")))
+  }
+}
+
+abstract class ConnectedComponents(log: Slf4jLogger) {
+  def getHashtables(): List[Byte]
+
+  def getHashValues(hashtable: Byte): Iterable[FileHash]
+
+  def makeBuckets(): List[List[Int]] = {
```

Comment:
It may be a good idea to add some [ScalaDoc](http://twitter.github.io/effectivescala/#Formatting-Comments) with very high-level description of the steps of the algorithm, so the reader has some idea of what is being done below.